pkg/streamingpromql/operators/functions/range_vectors.go

```diff
@@ -815,3 +816,82 @@ func quantileOverTime(step *types.RangeVectorStepData, _ float64, args []types.S
+	StepFunc:                       doubleExponentialSmoothing,
+	NeedsSeriesNamesForAnnotations: true,
+}
+
+// doubleExponentialSmoothing is similar to a weighted moving average, where
+// historical data has exponentially less influence on the current data. It also
+// accounts for trends in data. The smoothing factor (0 < sf < 1) affects how
+// historical data will affect the current data. A lower smoothing factor
+// increases the influence of historical data. The trend factor (0 < tf < 1)
+// affects how trends in historical data will affect the current data. A higher
+// trend factor increases the influence. of trends. Algorithm taken from
+// https://en.wikipedia.org/wiki/Exponential_smoothing .
+func doubleExponentialSmoothing(step *types.RangeVectorStepData, _ float64, args []types.ScalarData, timeRange types.QueryTimeRange, emitAnnotation types.EmitAnnotationFunc, _ *limiting.MemoryConsumptionTracker) (float64, bool, *histogram.FloatHistogram, error) {
+	smoothingFactor := args[0].Samples[timeRange.PointIndex(step.StepT)].F
+	if smoothingFactor <= 0 || smoothingFactor >= 1 {
+		return 0, false, nil, fmt.Errorf("invalid smoothing factor. Expected: 0 < sf < 1, got: %f", smoothingFactor)
+	}
+
+	trendFactor := args[1].Samples[timeRange.PointIndex(step.StepT)].F
+	if trendFactor <= 0 || trendFactor >= 1 {
+		return 0, false, nil, fmt.Errorf("invalid trend factor. Expected: 0 < tf < 1, got: %f", trendFactor)
+	}
+
+	fHead, fTail := step.Floats.UnsafePoints()
+
+	if step.Floats.Any() && step.Histograms.Any() {
+		emitAnnotation(annotations.NewHistogramIgnoredInMixedRangeInfo)
+	}
+
+	return runDoubleExponentialSmoothing(fHead, fTail, smoothingFactor, trendFactor)
+}
+
+func runDoubleExponentialSmoothing(fHead []promql.FPoint, fTail []promql.FPoint, smoothingFactor float64, trendFactor float64) (float64, bool, *histogram.FloatHistogram, error) {
+	if (len(fHead) + len(fTail)) < 2 {
+		return 0, false, nil, nil
+	}
+
+	var smooth0, smooth1, estimatedTrend float64
+
+	// Initial value of smooth1 is the first sample.
+	smooth1 = fHead[0].F
+
+	// Initial estimated trend is the difference between the first and second samples.
+	// First sample will always be in the head.
+	// If we have only one point in head, the second sample is the first index in tail.
+	if len(fHead) == 1 {
+		estimatedTrend = fTail[0].F - fHead[0].F
+	} else {
+		estimatedTrend = fHead[1].F - fHead[0].F
+	}
+
+	actualPointIndex := 0
+	accumulate := func(samples []promql.FPoint) {
+		var x, y float64
+
+		for _, sample := range samples {
+			// Scale the raw value against the smoothing factor.
+			x = smoothingFactor * sample.F
+
+			// Scale the last smoothed value for all samples except the first two points (point at actualPointIndex 0 and 1),
+			// regardless whether they are in head or tail.
+			if actualPointIndex > 1 {
+				estimatedTrend = trendFactor*(smooth1-smooth0) + (1-trendFactor)*estimatedTrend
+			} else {
+				actualPointIndex += 1
```

Comment:
[nit]

```suggestion
				actualPointIndex++
```

Reply:
Why not use a bool for actualPointIndex? Maybe for clarity it could be `firstPointAfterInit bool`

Reply:
Initially I was thinking to track the first two steps, but apparently we only need to skip the first iteration, hence boolean works. Updated in 19ad21cc120d87578d98598efdfc5bda7e2aa5a3